### ISSUE TITLE
Add support for validation message

### ISSUE THREAD

**Issue by ioquatix** [teamcapybara/capybara#2363]
With HTML5 form validations becoming more common, it might be nice to add support for validation messages:

The code currently looks like this:

```ruby
message = page.find("input[name=terms_of_service]").native.attribute("validationMessage")
expect(message).to be =~ /Please (check|tick) this box/
```

Maybe we can add a helper for this:

```ruby
message = validation_message(from: 'terms_of_service')
expect(message).to be =~ /Please (check|tick) this box/
```

This seems like a commonly asked question, and the given solution seems a bit verbose/fragile so could be great to provide a helper.

**Comment by twalpole** [comment]
Where has this been a commonly asked question, and who is suggesting that code as a solution (stop listening to them)?

1. why call `.native` - that's a sure sign you're doing something wrong. Instead to get the message just use the normal capybara methods (and stop using generic `find` when you are looking for supported types of elements)

    page.find_field('terms_of_service')['validationMessage']

2. you should never use basic RSpec matchers (eq, be, etc) with anything Capybara related - there's always a better way that allows for maintaining the benefits of Capybaras retrying behavior.  In this case if you wanted to verify the message on a specific element you could use a filter block

    expect(page).to have_field('terms_of_service) { |f|  f['validationMessage'] =~ /Please (check|tick) this box/ }

3. If you want to make that simpler in your project you should either add a custom selector, or probably better just extend the existing `field` selector with a new filter

        Capybara.modify_selector(:field) do
          node_filter(:validationMessage) { |node, msg|  node['validationMessage'] =~ msg }
        end

    which would allow you to do

        expect(page).to have_field('terms_of_service', validation_message: /Please (check|tick) this box/)

**Comment by ioquatix** [comment]
- https://www.google.com/search?q=capybara+form+validation+message
- https://medium.com/@tmikeschu/testing-html5-messages-with-rspec-selenium-5f5834c5af10
- https://stackoverflow.com/questions/6114385/capybara-webkit-how-to-test-client-side-validations-required-input-elemen

And tons of other pages about it.

I'll update stack overflow with your recommended answer, but I suggest this might need some more documentation as people are obviously doing the wrong thing and that approach has been entrenched by popularised wrong answers.

**Comment by twalpole** [comment]
Yes there are a number of old tutorials which give bad advice.  MAYBE they were the best that could be done when they were written, but things move on - and unfortunately there's no way to get old posts/tutorials removed/updated (I've tried on multiple occasions, and since given up) and people don't seem to look at posts from 3 years ago and think "hmmm, maybe thinks are different now".  As for `more documentation`  PRs are welcome.  However 95+% of people won't read documentation, they want to be spoon fed ready to copy/paste answers, and there's only so many hours in the day to deal with that.

**Comment by ioquatix** [comment]
> Yes there are a number of old tutorials which give bad advice. MAYBE they were the best that could be done when they were written, but things move on - and unfortunately there's no way to get old posts/tutorials removed/updated (I've tried on multiple occasions, and since given up) and people don't seem to look at posts from 3 years ago and think "hmmm, maybe thinks are different now".

Actually, that's exactly why I made the issue. I literally stated in the original post: "This seems like a commonly asked question, and the given solution seems a bit verbose/fragile so could be great to provide a helper."

> PRs are welcome. However 95+% of people won't read documentation, they want to be spoon fed ready to copy/paste answers, and there's only so many hours in the day to deal with that.

I actually did spend a chunk of time trying to figure out the best solution and that's why I ended up here making an issue trying to actually get a better solution, because I searched through the source code and could not find any reference to `validationMessage` so assumed it was not a supported/documented use case.

It's not obvious how one is supposed to find content which is not actual HTML content, I tried the obvious thing of expecting the page to have the content of the error message, etc.

**Comment by ioquatix** [comment]
https://github.com/teamcapybara/capybara/commit/144ecd93c9095a5d4bdce0b254977d60e5da73bc

**Comment by ioquatix** [comment]
Can you tag this with the released version (e.g. milestone) so I know when to update my own code? Thanks :)

**Comment by twalpole** [comment]
It doesn’t have a release version yet - if it makes it in the next release it’ll be documented in the release notes

**Comment by ioquatix** [comment]
Awesome! I find it's really helpful to add it as a "Milestone" on the issue.

**Comment by ioquatix** [comment]
Looks like this was released in v3.33.0 - great.
